### PR TITLE
Backend specific start sequence

### PR DESCRIPTION
**Pull request by criezy**
This pull request simplify creating game bundles on macOS.

It takes two of the recent additions and adapt them to be more platform-aware, with macOS custom implementations.
 - PR #4586
 - Auto-run mode added in commits 6cb1c7d and 399285b.

It also makes some adjustment to the macOS backend to better support game bundles.

Most of the changes here should also be easily transferable to the iOS backend. And other backends could also possibly have a different customised experience.

In details:
1. It introduces a new `updateStartSettings()` method in `OSystem` that is called after command line parameters have been parsed, but before processing them. This way backends can customise the start command and settings.
2. The auto-run code is moved to the default implementation of this new method.
3. On macOS, when running from a bundle, this default implementation is replaced with a custom implementation that looks for a `game` folder in the bundle resource directory. If such a folder is found the game is auto detected and run.
4. On macOS, it also looks for a `scummvm.ini` in the bundle resource directory, and if found it is used as initial config file.
5. Finally on macOS the default config and log file names now use the bundle name. By default (bundle name is `ScummVM`) the same name as before are used. But if create a game bundle using the scummvm executable, and use the game name as app bundle, it will now use its own separate config and log files.

That means that you can now create a game bundle with the following steps:
1. Copy the `ScummVM.app` bundle for example to `MyGame.app`
2. Create a `MyGame.app/Contents/Resources/game/` folder and copy your game data file to that folder.
3. Edit the bundle name and bundle display name in the `MyGame.app/Contents/Info.plist` file.
4. Create a new icons file in `MyGame.app/Contents/Resources/` and edit the icon file in the `MyGame.app/Contents/Info.plist` file.
5. Create a `MyGame.app/Contents/Resources/scummvm.ini` file with the default settings for your game bundle (for example to set fullscreen to true).

Step 3, 4 and 5 are optional but allow to create a proper game bundle with its own identity.

I am planning to create a small swift application to perform those steps (that will provide me with a good opportunity to dust off my swift skills).

### CODE REVIEW

**Review by sev- — APPROVED** [review]
Looks good to me, including the ISystem refactoring

File: backends/platform/sdl/macosx/macosx_osys_misc.mm [hunk 1]
```diff
@@ -98,6 +98,24 @@
 		command = "auto-detect";
 		settings["path"] = [gamePath fileSystemRepresentation];
 	}
+
+	// Else if the bundle contains a games directory, add them to the launcher
+	// Should we only do it on first start (if the config file does not yet exist) to avoid adding
+	// backe games removed by the user? That would also speed up subsequent starts.
```
Comment (raziel-) [inline backends/platform/sdl/macosx/macosx_osys_misc.mm:104]: typo?

avoid adding "back" games?

File: backends/platform/sdl/macosx/macosx_osys_misc.mm [hunk 2]
```diff
@@ -98,6 +98,24 @@
...
+	// However doing it always, means that adding more games in the bundle  in the future will
+	// properly add them to the launcher.
+	else {
+		NSString *gamesPath = [[bundle resourcePath] stringByAppendingPathComponent:@"games"];
+		isDir = false;
+		exists = [[NSFileManager defaultManager] fileExistsAtPath:gamesPath isDirectory:&isDir];
+		if (exists && isDir) {
+			// Ause auto-detection
```
Comment (raziel-) [inline backends/platform/sdl/macosx/macosx_osys_misc.mm:112]: typo?

"Use" auto-detection

File: base/commandLine.cpp [hunk 3]
```diff
@@ -1707,48 +1712,51 @@ bool processSettings(Common::String &command, Common::StringMap &settings, Commo
 		}
 	}
 
+	// For command that normally exit, check if --no-exit was specified
```
Comment (raziel-) [inline base/commandLine.cpp:1715]: typo?

For "commands" that normally exit,...

File: backends/platform/ios7/ios7_osys_misc.mm [hunk 4]
```diff
@@ -37,6 +45,74 @@ static inline void execute_on_main_thread_async(void (^block)(void)) {
...
+		// Check the last version stored in the user settings.
+		NSString *versionString = [NSString stringWithUTF8String:gScummVMFullVersion];
+		NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
+		NSString *lastVersion = [defaults stringForKey:@"lastVersion"];
+		if (lastVersion && [lastVersion isEqualToString:versionString])
+			return;
+		[defaults setObject:versionString forKey:@"lastVersion"];
+
+		// If the bundle contains a games directory, add them to the launcher
+		NSString *gamesPath = [[bundle resourcePath] stringByAppendingPathComponent:@"games"];
```
Comment (sev-) [inline backends/platform/ios7/ios7_osys_misc.mm:102]: Will this detection run every time at the scummvm startup?

File: backends/platform/sdl/macosx/macosx_osys_misc.mm [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+		}
+		autorun.close();
+	}
+
+	// If the bundle contains a game directory, auto-detect it
+	NSString *gamePath = [[bundle resourcePath] stringByAppendingPathComponent:@"game"];
+	BOOL isDir = false;
+	BOOL exists = [[NSFileManager defaultManager] fileExistsAtPath:gamePath isDirectory:&isDir];
+	if (exists && isDir) {
+		// Ause auto-detection
```
Comment (sev-) [inline backends/platform/sdl/macosx/macosx_osys_misc.mm:99]: A typo?

File: backends/platform/sdl/macosx/macosx_osys_misc.mm [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+	// The rest of the function has some commands executed only the first time after each version change
+	// Check the last version stored in the user settings.
+	NSString *versionString = [NSString stringWithUTF8String:gScummVMFullVersion];
+	NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
+	NSString *lastVersion = [defaults stringForKey:@"lastVersion"];
+	if (lastVersion && [lastVersion isEqualToString:versionString])
+		return;
+	[defaults setObject:versionString forKey:@"lastVersion"];
+
+	// If the bundle contains a games directory, add them to the launcher
```
Comment (sev-) [inline backends/platform/sdl/macosx/macosx_osys_misc.mm:114]: Does it make sense to perhaps share this code between ios/macosx?

File: base/commandLine.cpp [hunk 7]
```diff
@@ -1917,10 +1925,27 @@ bool processSettings(Common::String &command, Common::StringMap &settings, Commo
 		nullptr
 	};
 
+	// Skip some settings that should only be used for the command-line commands
+	static const char * const skipSettings[] = {
```
Comment (sev-) [inline base/commandLine.cpp:1929]: I wonder, how we can make sure that in the future if such commands are added, they get into this list as well...

**Review by sev- — APPROVED** [review]
Have couple of questions so far, but generally, looks good to me.

File: backends/platform/ios7/ios7_osys_misc.mm (hunk 4 above)
Comment (criezy) [reply backends/platform/ios7/ios7_osys_misc.mm:106]: No. Commit 5a6fbc9 ensures this is only run the first time after the ScummVM version changes (so on first install and each subsequent upgrade).

File: backends/platform/sdl/macosx/macosx_osys_misc.mm (hunk 6 above)
Comment (criezy) [reply backends/platform/sdl/macosx/macosx_osys_misc.mm:114]: It probably does. The code is almost identical, with the only differences coming in the last commit for the sandboxed file paths.

File: backends/platform/sdl/macosx/macosx_osys_misc.mm (hunk 5 above)
Comment (criezy) [reply backends/platform/sdl/macosx/macosx_osys_misc.mm]: Fixed

**Comment by criezy** [comment]
This pull request has reach a point where it works on both iOS and macOS.
However there is some code duplication between the two backends that could possibly be factored.

Also the change to the sandboxed file system factory and fs nodes, to allow running games directly from the bundle, needs to be reviewed. This is only used by the iOS backend, but I am not sure I properly considered all the implications of that change.

**Comment by sev-** [comment]
Merging.

**Pull request merged**